src/Avalonia.Visuals/Media/Transformation/TransformOperation.cs

```diff
@@ -0,0 +1,230 @@
+
+                    result.Data.Rotate.Angle = InterpolationUtilities.InterpolateScalars(fromAngle, toAngle, progress);
+
+                    result.Bake();
+
+                    break;
+                }
+                case OperationType.Scale:
+                {
+                    double fromX = fromIdentity ? 1 : fromValue.Data.Scale.X;
+                    double fromY = fromIdentity ? 1 : fromValue.Data.Scale.Y;
+
+                    double toX = toIdentity ? 1 : toValue.Data.Scale.X;
+                    double toY = toIdentity ? 1 : toValue.Data.Scale.Y;
+
+                    result.Data.Scale.X = InterpolationUtilities.InterpolateScalars(fromX, toX, progress);
+                    result.Data.Scale.Y = InterpolationUtilities.InterpolateScalars(fromY, toY, progress);
+
+                    result.Bake();
+
+                    break;
+                }
+                case OperationType.Skew:
+                {
+                    double fromX = fromIdentity ? 0 : fromValue.Data.Skew.X;
+                    double fromY = fromIdentity ? 0 : fromValue.Data.Skew.Y;
+
+                    double toX = toIdentity ? 0 : toValue.Data.Skew.X;
+                    double toY = toIdentity ? 0 : toValue.Data.Skew.Y;
+
+                    result.Data.Skew.X = InterpolationUtilities.InterpolateScalars(fromX, toX, progress);
+                    result.Data.Skew.Y = InterpolationUtilities.InterpolateScalars(fromY, toY, progress);
+
+                    result.Bake();
+
+                    break;
+                }
+                case OperationType.Matrix:
+                {
+                    var fromMatrix = fromIdentity ? Matrix.Identity : fromValue.Matrix;
+                    var toMatrix = toIdentity ? Matrix.Identity : toValue.Matrix;
+
+                    if (!Matrix.TryDecomposeTransform(fromMatrix, out Matrix.Decomposed fromDecomposed) ||
+                        !Matrix.TryDecomposeTransform(toMatrix, out Matrix.Decomposed toDecomposed))
+                    {
+                        return false;
+                    }
+
+                    var interpolated =
+                        InterpolationUtilities.InterpolateDecomposedTransforms(
+                            ref fromDecomposed, ref toDecomposed,
+                            progress);
+
+                    result.Matrix = InterpolationUtilities.ComposeTransform(interpolated);
+
+                    break;
+                }
+                case OperationType.Identity:
+                {
+                    // Do nothing.
+                    break;
+                }
+            }
+
+            return true;
+        }
+
+        private static bool IsOperationIdentity(ref TransformOperation? operation)
+        {
+            return !operation.HasValue || operation.Value.IsIdentity;
+        }
+
+        [StructLayout(LayoutKind.Explicit)]
+        public struct DataLayout
+        {
+            [FieldOffset(0)] public SkewLayout Skew;
```

Comment:
These all have the same offset, does that mean they will overwrite each other?

Reply:
@MarchingCube reported, this is intended, only 1 operation kind at a time is required, and this saves memory. C Union functionality intended.